src/btf.cpp

```diff
@@ -727,6 +727,91 @@ FuncParamLists BTF::get_params_from_btf(
   return params;
 }
 
+FuncParamLists BTF::get_kprobes_params_from_btf(
+    const BTFObj &btf_obj,
+    const std::set<std::string> &funcs) const
+{
+  std::stringstream type;
+  auto *dump = dump_new(btf_obj.btf, dump_printf, &type);
+  if (auto err = libbpf_get_error(dump)) {
+    char err_buf[256] = {};
+    libbpf_strerror(err, err_buf, sizeof(err_buf));
+    LOG(ERROR) << "BTF: failed to initialize dump (" << err_buf << ")";
+    return {};
+  }
+  SCOPE_EXIT
+  {
+    btf_dump__free(dump);
+  };
+
+  FuncParamLists params;
+  auto id = start_id(btf_obj.btf), max = type_cnt(btf_obj.btf);
+  for (; id <= max; id++) {
+    const struct btf_type *t = btf__type_by_id(btf_obj.btf, id);
+    if (!t)
+      continue;
+
+    if (!btf_is_func(t))
+      continue;
+
+    std::string func_name;
+    const std::string pure_func_name = btf__name_by_offset(btf_obj.btf,
+                                                           t->name_off);
+    const std::string obj_func_name = btf_obj.name + ":" + pure_func_name;
+
+    // First match the module prefix name, then match the pure function name.
+    // For example, first match "kprobe:vmlinux:do_sys*", then "kprobe:do_sys*".
+    // The same goes for other modules, such as "kvm".
+    if (funcs.contains(obj_func_name))
+      func_name = obj_func_name;
+    else if (funcs.contains(pure_func_name))
+      func_name = pure_func_name;
+    else
+      continue;
+
+    t = btf__type_by_id(btf_obj.btf, t->type);
+    if (!t)
+      continue;
+
+    BPFTRACE_LIBBPF_OPTS(btf_dump_emit_type_decl_opts,
+                         decl_opts,
+                         .field_name = "");
+
+    const auto *p = btf_params(t);
+    for (__u16 j = 0, len = btf_vlen(t); j < len; j++, p++) {
+      const char *arg_name = btf__name_by_offset(btf_obj.btf, p->name_off);
+
+      // set by dump_printf callback
+      type.str("");
+      if (btf_dump__emit_type_decl(dump, p->type, &decl_opts)) {
+        LOG(ERROR) << "failed to dump argument: " << arg_name;
+        break;
+      }
+
+      params[func_name].push_back(type.str() + " " + arg_name);
+    }
+
+    if (!t->type) {
+      params[func_name].push_back("void");
+      continue;
+    }
+
+    // set by dump_printf callback
+    type.str("");
+    if (btf_dump__emit_type_decl(dump, t->type, &decl_opts)) {
+      LOG(ERROR) << "failed to dump return type for: " << func_name;
+      break;
+    }
+
+    params[func_name].push_back(type.str());
```

Comment:
Not sure how clever we should get here but I think we know of some types that typically won't be in registers and therefore not accessible via the `argN` builtins e.g. double/float

https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#kprobe-and-kretprobe

Reply:
I thinks you're right, done, please review.